File: src/ISA/ISA.Spreadsheet/InvestigationFile/Protocols.fs
namespace ISA.Spreadsheet

open ISA
open ISA.API
open Comment
open Remark
open System.Collections.Generic

module Protocols = 

    let nameLabel = "Name"
    let protocolTypeLabel = "Type"
    let typeTermAccessionNumberLabel = "Type Term Accession Number"
    let typeTermSourceREFLabel = "Type Term Source REF"
    let descriptionLabel = "Description"
    let uriLabel = "URI"
    let versionLabel = "Version"
    let parametersNameLabel = "Parameters Name"
    let parametersTermAccessionNumberLabel = "Parameters Term Accession Number"
    let parametersTermSourceREFLabel = "Parameters Term Source REF"
    let componentsNameLabel = "Components Name"
    let componentsTypeLabel = "Components Type"
    let componentsTypeTermAccessionNumberLabel = "Components Type Term Accession Number"
    let componentsTypeTermSourceREFLabel = "Components Type Term Source REF"

    let labels = 
        [
            nameLabel;protocolTypeLabel;typeTermAccessionNumberLabel;typeTermSourceREFLabel;descriptionLabel;uriLabel;versionLabel;
            parametersNameLabel;parametersTermAccessionNumberLabel;parametersTermSourceREFLabel;
            componentsNameLabel;componentsTypeLabel;componentsTypeTermAccessionNumberLabel;componentsTypeTermSourceREFLabel
        ]

    let fromString name protocolType typeTermAccessionNumber typeTermSourceREF description uri version parametersName parametersTermAccessionNumber parametersTermSourceREF componentsName componentsType componentsTypeTermAccessionNumber componentsTypeTermSourceREF comments =
        let protocolType = OntologyAnnotation.fromString protocolType typeTermSourceREF typeTermAccessionNumber
        let parameters = ProtocolParameter.fromAggregatedStrings ';' parametersName parametersTermSourceREF parametersTermAccessionNumber
        let components = Component.fromAggregatedStrings ';' componentsName componentsType componentsTypeTermSourceREF componentsTypeTermAccessionNumber
        
        Protocol.make 
            None 
            (Option.fromValueWithDefault "" name |> Option.map URI.fromString) 
            (Option.fromValueWithDefault OntologyAnnotation.empty protocolType)
            (Option.fromValueWithDefault "" description)
            (Option.fromValueWithDefault "" uri |> Option.map URI.fromString) 
            (Option.fromValueWithDefault "" version)
            (Option.fromValueWithDefault [] parameters)
            (Option.fromValueWithDefault [] components) 
            (Option.fromValueWithDefault [] comments)


    let fromSparseTable (matrix : SparseTable) =
        
        List.init matrix.Length (fun i -> 

            let comments = 
                matrix.CommentKeys 
                |> List.map (fun k -> 
                    Comment.fromString k (matrix.TryGetValueDefault("",(k,i))))

            fromString
                (matrix.TryGetValueDefault("",(nameLabel,i)))
                (matrix.TryGetValueDefault("",(protocolTypeLabel,i)))
                (matrix.TryGetValueDefault("",(typeTermAccessionNumberLabel,i)))
                (matrix.TryGetValueDefault("",(typeTermSourceREFLabel,i)))
                (matrix.TryGetValueDefault("",(descriptionLabel,i)))
                (matrix.TryGetValueDefault("",(uriLabel,i)))
                (matrix.TryGetValueDefault("",(versionLabel,i)))
                (matrix.TryGetValueDefault("",(parametersNameLabel,i)))
                (matrix.TryGetValueDefault("",(parametersTermAccessionNumberLabel,i)))
                (matrix.TryGetValueDefault("",(parametersTermSourceREFLabel,i)))
                (matrix.TryGetValueDefault("",(componentsNameLabel,i)))
                (matrix.TryGetValueDefault("",(componentsTypeLabel,i)))
                (matrix.TryGetValueDefault("",(componentsTypeTermAccessionNumberLabel,i)))
                (matrix.TryGetValueDefault("",(componentsTypeTermSourceREFLabel,i)))
                comments
        )
    
    let toSparseTable (protocols: Protocol list) =
        let matrix = SparseTable.Create (keys = labels,length=protocols.Length + 1)
        let mutable commentKeys = []
        protocols
        |> List.iteri (fun i p ->
            let i = i + 1
            let protocolType,protocolSource,protocolAccession = p.ProtocolType |> Option.defaultValue OntologyAnnotation.empty |> OntologyAnnotation.toString 
            let parameterType,parameterSource,parameterAccession = p.Parameters |> Option.defaultValue [] |> ProtocolParameter.toAggregatedStrings ';' 
            let componentName,componentType,componentSource,componentAccession = p.Components |> Option.defaultValue [] |> Component.toAggregatedStrings ';' 

            do matrix.Matrix.Add ((nameLabel,i),                                (Option.defaultValue "" p.Name))
            do matrix.Matrix.Add ((protocolTypeLabel,i),                        protocolType)
            do matrix.Matrix.Add ((typeTermAccessionNumberLabel,i),             protocolAccession)
            do matrix.Matrix.Add ((typeTermSourceREFLabel,i),                   protocolSource)
            do matrix.Matrix.Add ((descriptionLabel,i),                         (Option.defaultValue "" p.Description))
            do matrix.Matrix.Add ((uriLabel,i),                                 (Option.defaultValue "" p.Uri))
            do matrix.Matrix.Add ((versionLabel,i),                             (Option.defaultValue "" p.Version))
            do matrix.Matrix.Add ((parametersNameLabel,i),                      parameterType)
            do matrix.Matrix.Add ((parametersTermAccessionNumberLabel,i),       parameterAccession)
            do matrix.Matrix.Add ((parametersTermSourceREFLabel,i),             parameterSource)
            do matrix.Matrix.Add ((componentsNameLabel,i),                      componentName)
            do matrix.Matrix.Add ((componentsTypeLabel,i),                      componentType)
            do matrix.Matrix.Add ((componentsTypeTermAccessionNumberLabel,i),   componentAccession)
            do matrix.Matrix.Add ((componentsTypeTermSourceREFLabel,i),         componentSource)

            match p.Comments with 
            | None -> ()
            | Some c ->
                c
                |> List.iter (fun comment -> 
                    let n,v = comment |> Comment.toString
                    commentKeys <- n :: commentKeys
                    matrix.Matrix.Add((n,i),v)
                )    
        )
        {matrix with CommentKeys = commentKeys |> List.distinct |> List.rev} 

    let fromRows (prefix : string option) lineNumber (rows : IEnumerator<SparseRow>) =
        match prefix with
        | Some p -> SparseTable.FromRows(rows,labels,lineNumber,p)
        | None -> SparseTable.FromRows(rows,labels,lineNumber)
        |> fun (s,ln,rs,sm) -> (s,ln,rs, fromSparseTable sm)
  
    let toRows prefix (protocols : Protocol list) =
        protocols
        |> toSparseTable
        |> fun m -> 
            match prefix with 
            | Some prefix -> SparseTable.ToRows(m,prefix)
            | None -> SparseTable.ToRows(m)

File: js/ISA/ISA.Spreadsheet/InvestigationFile/Protocols.js
import { reverse, cons, iterate, iterateIndexed, map as map_1, initialize, singleton, length, empty, ofArray } from "../../../fable_modules/fable-library.4.1.4/List.js";
import { OntologyAnnotation_toString_473B9D79, OntologyAnnotation_get_empty, OntologyAnnotation_fromString_Z7D8EB286 } from "../../ISA/JsonTypes/OntologyAnnotation.js";
import { Component_toAggregatedStrings, ProtocolParameter_toAggregatedStrings, Option_fromValueWithDefault, Component_fromAggregatedStrings, ProtocolParameter_fromAggregatedStrings } from "../Conversions.js";
import { Protocol_create_Z4D717767, Protocol_make } from "../../ISA/JsonTypes/Protocol.js";
import { defaultArg, map } from "../../../fable_modules/fable-library.4.1.4/Option.js";
import { URIModule_fromString } from "../../ISA/JsonTypes/URI.js";
import { SparseTable_ToRows_584133C0, SparseTable_FromRows_Z5579EC29, SparseTable, SparseTable_Create_Z2192E64B, SparseTable__TryGetValue_11FD62A8, SparseTable__TryGetValueDefault_5BAE6133, SparseTable_GetEmptyComments_Z15A4F148 } from "../SparseTable.js";
import { Comment_toString, Comment_fromString } from "../Comment.js";
import { addToDict } from "../../../fable_modules/fable-library.4.1.4/MapUtil.js";
import { List_distinct } from "../../../fable_modules/fable-library.4.1.4/Seq2.js";
import { stringHash } from "../../../fable_modules/fable-library.4.1.4/Util.js";

export const nameLabel = "Name";

export const protocolTypeLabel = "Type";

export const typeTermAccessionNumberLabel = "Type Term Accession Number";

export const typeTermSourceREFLabel = "Type Term Source REF";

export const descriptionLabel = "Description";

export const uriLabel = "URI";

export const versionLabel = "Version";

export const parametersNameLabel = "Parameters Name";

export const parametersTermAccessionNumberLabel = "Parameters Term Accession Number";

export const parametersTermSourceREFLabel = "Parameters Term Source REF";

export const componentsNameLabel = "Components Name";

export const componentsTypeLabel = "Components Type";

export const componentsTypeTermAccessionNumberLabel = "Components Type Term Accession Number";

export const componentsTypeTermSourceREFLabel = "Components Type Term Source REF";

export const labels = ofArray([nameLabel, protocolTypeLabel, typeTermAccessionNumberLabel, typeTermSourceREFLabel, descriptionLabel, uriLabel, versionLabel, parametersNameLabel, parametersTermAccessionNumberLabel, parametersTermSourceREFLabel, componentsNameLabel, componentsTypeLabel, componentsTypeTermAccessionNumberLabel, componentsTypeTermSourceREFLabel]);

export function fromString(name, protocolType, typeTermAccessionNumber, typeTermSourceREF, description, uri, version, parametersName, parametersTermAccessionNumber, parametersTermSourceREF, componentsName, componentsType, componentsTypeTermAccessionNumber, componentsTypeTermSourceREF, comments) {
    const protocolType_1 = OntologyAnnotation_fromString_Z7D8EB286(protocolType, typeTermSourceREF, typeTermAccessionNumber);
    const parameters = ProtocolParameter_fromAggregatedStrings(";", parametersName, parametersTermSourceREF, parametersTermAccessionNumber);
    const components = Component_fromAggregatedStrings(";", componentsName, componentsType, componentsTypeTermSourceREF, componentsTypeTermAccessionNumber);
    return Protocol_make(void 0, map(URIModule_fromString, Option_fromValueWithDefault("", name)), Option_fromValueWithDefault(OntologyAnnotation_get_empty(), protocolType_1), Option_fromValueWithDefault("", description), map(URIModule_fromString, Option_fromValueWithDefault("", uri)), Option_fromValueWithDefault("", version), Option_fromValueWithDefault(empty(), parameters), Option_fromValueWithDefault(empty(), components), Option_fromValueWithDefault(empty(), comments));
}

export function fromSparseTable(matrix) {
    if ((matrix.ColumnCount === 0) && (length(matrix.CommentKeys) !== 0)) {
        return singleton(Protocol_create_Z4D717767(void 0, void 0, void 0, void 0, void 0, void 0, void 0, void 0, SparseTable_GetEmptyComments_Z15A4F148(matrix)));
    }
    else {
        return initialize(matrix.ColumnCount, (i) => {
            const comments_1 = map_1((k) => Comment_fromString(k, SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [k, i])), matrix.CommentKeys);
            return fromString(SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [nameLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [protocolTypeLabel, i]), SparseTable__TryGetValue_11FD62A8(matrix, [typeTermAccessionNumberLabel, i]), SparseTable__TryGetValue_11FD62A8(matrix, [typeTermSourceREFLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [descriptionLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [uriLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [versionLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [parametersNameLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [parametersTermAccessionNumberLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [parametersTermSourceREFLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [componentsNameLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [componentsTypeLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [componentsTypeTermAccessionNumberLabel, i]), SparseTable__TryGetValueDefault_5BAE6133(matrix, "", [componentsTypeTermSourceREFLabel, i]), comments_1);
        });
    }
}

export function toSparseTable(protocols) {
    const matrix = SparseTable_Create_Z2192E64B(void 0, labels, void 0, length(protocols) + 1);
    let commentKeys = empty();
    iterateIndexed((i, p) => {
        const i_1 = (i + 1) | 0;
        const pt_1 = OntologyAnnotation_toString_473B9D79(defaultArg(p.ProtocolType, OntologyAnnotation_get_empty()), true);
        const pAgg = ProtocolParameter_toAggregatedStrings(";", defaultArg(p.Parameters, empty()));
        const cAgg = Component_toAggregatedStrings(";", defaultArg(p.Components, empty()));
        addToDict(matrix.Matrix, [nameLabel, i_1], defaultArg(p.Name, ""));
        addToDict(matrix.Matrix, [protocolTypeLabel, i_1], pt_1.TermName);
        addToDict(matrix.Matrix, [typeTermAccessionNumberLabel, i_1], pt_1.TermAccessionNumber);
        addToDict(matrix.Matrix, [typeTermSourceREFLabel, i_1], pt_1.TermSourceREF);
        addToDict(matrix.Matrix, [descriptionLabel, i_1], defaultArg(p.Description, ""));
        addToDict(matrix.Matrix, [uriLabel, i_1], defaultArg(p.Uri, ""));
        addToDict(matrix.Matrix, [versionLabel, i_1], defaultArg(p.Version, ""));
        addToDict(matrix.Matrix, [parametersNameLabel, i_1], pAgg.TermNameAgg);
        addToDict(matrix.Matrix, [parametersTermAccessionNumberLabel, i_1], pAgg.TermAccessionNumberAgg);
        addToDict(matrix.Matrix, [parametersTermSourceREFLabel, i_1], pAgg.TermSourceREFAgg);
        addToDict(matrix.Matrix, [componentsNameLabel, i_1], cAgg.NameAgg);
        addToDict(matrix.Matrix, [componentsTypeLabel, i_1], cAgg.TermNameAgg);
        addToDict(matrix.Matrix, [componentsTypeTermAccessionNumberLabel, i_1], cAgg.TermAccessionNumberAgg);
        addToDict(matrix.Matrix, [componentsTypeTermSourceREFLabel, i_1], cAgg.TermSourceREFAgg);
        const matchValue = p.Comments;
        if (matchValue != null) {
            iterate((comment) => {
                const patternInput = Comment_toString(comment);
                const n = patternInput[0];
                commentKeys = cons(n, commentKeys);
                addToDict(matrix.Matrix, [n, i_1], patternInput[1]);
            }, matchValue);
        }
    }, protocols);
    return new SparseTable(matrix.Matrix, matrix.Keys, reverse(List_distinct(commentKeys, {
        Equals: (x, y) => (x === y),
        GetHashCode: stringHash,
    })), matrix.ColumnCount);
}

export function fromRows(prefix, lineNumber, rows) {
    const tupledArg = (prefix == null) ? SparseTable_FromRows_Z5579EC29(rows, labels, lineNumber) : SparseTable_FromRows_Z5579EC29(rows, labels, lineNumber, prefix);
    return [tupledArg[0], tupledArg[1], tupledArg[2], fromSparseTable(tupledArg[3])];
}

export function toRows(prefix, protocols) {
    const m = toSparseTable(protocols);
    if (prefix == null) {
        return SparseTable_ToRows_584133C0(m);
    }
    else {
        return SparseTable_ToRows_584133C0(m, prefix);
    }
}

